lib/fog/compute/kubevirt.rb

```diff
@@ -84,26 +118,142 @@ def initialize(options={})
+        # @return [WatchWrapper] The watcher.
+        #
+        def watch_nodes(opts = {})
+          mapper = Proc.new do |notice|
+            node = OpenStruct.new(Node.parse(notice.object))
+            populate_notice_attributes(node, notice)
+            node
+          end
+          watch = kube_client.watch_nodes(opts)
+
+          WatchWrapper.new(watch, mapper)
         end
 
-        def kube_client
-          @kubeclient
+        #
+        # Returns a watcher for offline virtual machines.
+        #
+        # @param opts [Hash] A hash with options for the watcher.
+        # @return [WatchWrapper] The watcher.
+        #
+        def watch_offline_vms(opts = {})
+          mapper = Proc.new do |notice|
+            offlinevm = OpenStruct.new(Offlinevm.parse(notice.object))
+            populate_notice_attributes(offlinevm, notice)
+            offlinevm
+          end
+          watch = kubevirt_client.watch_offline_virtual_machines(opts)
+
+          WatchWrapper.new(watch, mapper)
+        end
+
+        #
+        # Returns a watcher for live virtual machines.
+        #
+        # @param opts [Hash] A hash with options for the watcher.
+        # @return [WatchWrapper] The watcher.
+        #
+        def watch_live_vms(opts = {})
+          mapper = Proc.new do |notice|
+            livevm = OpenStruct.new(Livevm.parse(notice.object))
+            populate_notice_attributes(livevm, notice)
+            livevm
+          end
+          watch = kubevirt_client.watch_virtual_machines(opts)
+
+          WatchWrapper.new(watch, mapper)
+        end
+
+        #
+        # Returns a watcher for templates.
+        #
+        # @param opts [Hash] A hash with options for the watcher.
+        # @return [WatchWrapper] The watcher.
+        #
+        def watch_templates(opts = {})
+          mapper = Proc.new do |notice|
+            template = OpenStruct.new(Template.parse(notice.object))
+            populate_notice_attributes(template, notice)
+            template
+          end
+          watch = openshift_client.watch_templates(opts)
+
+          WatchWrapper.new(watch, mapper)
+        end
+
+        #
+        # Calculates the URL of the SPICE proxy server.
+        #
+        # @return [String] The URL of the spice proxy server.
+        #
+        def spice_proxy_url
+          service = kube_client.get_service('spice-proxy', @namespace)
+          host = service.spec.externalIPs.first
+          port = service.spec.ports.first.port
+          url = URI::Generic.build(
+            :scheme => 'http',
+            :host   => host,
+            :port   => port,
+          )
+          url.to_s
+        end
+
+        private
+
+
+        #
+        # Populates required notice attributes
+        #
+        # @param object entity to populate
+        # @param notice the source of the data to populate from
+        def populate_notice_attributes(object, notice)
+          object.metadata = notice.object.metadata
```

Comment:
The metadata is being accessed from the runner.rb and from the refresh_memory.rb. 

However it can be refined to the actual used attributes (uid, resource_version, type, kind or others). 

I’ll work on reducing that blown object with minimal set of attributes.